### PR TITLE
[core] Vendor msgpack dependency

### PR DESCRIPTION
**Pull request by brettlangdon**
In this PR we are adding [msgpack](https://github.com/msgpack/msgpack-python) as a vendored dependency in `ddtrace/vendor/msgpack`.

This vendoring required copying over source files from `msgpack` and updating all imports from `from msgpack import *` to `from ddtrace.vendor.msgpack import *`.

As well, we needed to update `_packer.pyx` and `_unpacker.pyx` to import from the expected location and use `cython` to build and vendor `_cmsgpack.cpp`.

Luckily enough we already had pieces in place in `setup.py` to try and build a C-extension and back off to pure Python if it fails so there was not a lot of changes needed there.

### CODE REVIEW

**Comment by brettlangdon** [comment]
@jd yup!

We want to vendor all of our dependencies mostly so we don't mess with any dependencies used by our users, we can rely on specific packages/versions without having to lock our customers into those packages/versions as well.

Specifically for `msgpack-python`, there are two reasons why:

1. We were relying on `msgpack-python` which is now deprecated in favor of `msgpack` (just a PyPI package rename mostly), we were blocking our customers from updating their codebase as well.
2. Because we only depended on `msgpack-python` (no version constraint) it means our codebase would have to be compatible with all versions of it. We were explicitly relying on/validating the existence of an internal/C module which was renamed in a newer version of the library, which means our code doesn't work correctly for anyone using a newer version of `msgpack-python`

Now, the only solution to either of those problems isn't "vendor", but that is the direction we wanted to head in for all of our dependencies anyways. This is actually the last dependency of ours to get vendored.

File: setup.py [hunk 1]
```diff
@@ -85,9 +85,7 @@ def run_tests(self):
     long_description_content_type='text/markdown',
     license='BSD',
     packages=find_packages(exclude=['tests*']),
-    install_requires=[
-        'msgpack-python',
-    ],
+    install_requires=[],
```
Comment (jd) [inline setup.py]: Just remove the line entirely?

File: tox.ini [hunk 2]
```diff
@@ -82,7 +82,7 @@ envlist =
     mako_contrib-{py27,py34,py35,py36}-mako{010,100}
     molten_contrib-{py36}-molten{070,072}
     mongoengine_contrib-{py27,py34,py35,py36}-mongoengine{015}
-    msgpack_contrib-{py27,py34}-msgpack{03,04,05}
+    msgpack_contrib-{py27,py34}
```
Comment (jd) [inline tox.ini]: I'm confused; if we don't install any specific version of `msgpack`, which version it is tested against?
Are we removing support for 0.3, 0.4 and 0.5?

Comment (brettlangdon) [reply tox.ini]: This env/suite just tests our encoders (we have a JSON and msgpack encoder). So we are moving from testing the various versions the user could have installed to just our own internal one.

Let me rename this suite... I should also check if it is being run with the generic tracer suite or not, if so we can remove entirely.

Comment (jd) [reply tox.ini]: I agree the name is really confusing. :)

Comment (brettlangdon) [reply tox.ini]: I just removed this test env entirely since these tests get run by another anyways.

File: tox.ini [hunk 3]
```diff
@@ -128,7 +127,7 @@ deps =
 # distribution build.
     !ddtracerun: wrapt
     !msgpack03-!msgpack04-!msgpack05-!ddtracerun: msgpack-python
-    pytest>=3
+    pytest>=3.0.0,<4.0.0
```
Comment (jd) [inline tox.ini]: That looks wrong.

Comment (brettlangdon) [reply tox.ini]: yup, you are correct, I messed up the merge conflict I had, great catch!

Comment (brettlangdon) [reply tox.ini]:
```suggestion
    pytest>=3.0.0
```

File: tox.ini [hunk 4]
```diff
@@ -128,7 +127,7 @@ deps =
 # distribution build.
     !ddtracerun: wrapt
     !msgpack03-!msgpack04-!msgpack05-!ddtracerun: msgpack-python
-    pytest>=3
+    pytest>=3.0.0
```
Comment (brettlangdon) [inline tox.ini]:
```suggestion
    pytest>=3
```

**Pull request merged**